File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
n, c = map(int, input().split())
lst = []
for _ in range(n):
    lst.append(int(sys.stdin.readline()))
lst.sort()

start = 1
end = lst[-1] - lst[0]
ans = 0

while start <= end:
    mid = (start + end) // 2
    cnt = 1
    tmp = lst[0]

    for i in range(1,len(lst)):
        if lst[i] >= tmp + mid:
            cnt += 1
            tmp = lst[i]    
    if cnt >= c:
        start = mid + 1
        ans = mid
    else:
        end = mid - 1

print(ans)

File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


int main() {
	int n, c, ans;
	cin >> n >> c;
	vector<int> v(n);
	
	for (int i = 0; i < n; i++) cin >> v[i];
	sort(v.begin(), v.end());

	int mid;
	int left = 1, right = v[n-1]-v[0];
	while (left <= right) {
		mid = (left + right) / 2;
		int cnt = 1;
		int cur = v[0];
		for (int i = 1; i < n; i++) {
			if (v[i] - mid >= cur) {
				cnt++;
				cur = v[i];
			}
		}
		if (cnt >= c) {
			left = mid + 1;
			ans = mid;
		}
		else right = mid - 1;
	}
	cout << ans;
}